src/Uno.UI/UI/Xaml/Media/VisualTreeHelper.cs

```diff
@@ -543,14 +543,15 @@ internal static IReadOnlyList<_View> ClearChildren(UIElement view)
 
 			var isChildStale = isStale;
 
+			// We only take ZIndex into account on skia, which supports Canvas.Zindex for non-canvas panels.
+			// Once Canvas.ZIndex renders correctly elsewhere, remove the conditional OrderBy
+			// https://github.com/unoplatform/uno/issues/325
 			using var child = children
-#if __IOS__ || __MACOS__ || __ANDROID__ || IS_UNIT_TESTS
-				.Reverse().GetEnumerator();
-#else
-				// On Skia and Wasm, we can get concrete data structure (MaterializableList in this case) instead of IEnumerable<T>.
-				// It has an efficient "ReverseEnumerator". This will also avoid the boxing allocations of the enumerator when it's a struct.
-				.GetReverseEnumerator();
+#if __SKIA__
+				.OrderBy(e => e.Visual.ZIndex) // Equivalent to GetValue(Canvas.ZIndexProperty) on skia
```

Comment:
What about the other loop in `SearchDownForLeafCore`? Can we have a test case that shows the difference if we order by ZIndex in it? @dr1rrb may be able to help here too.

Reply:
I took it one step further and cached the reverse-sorted list in the MaterializableList, since we will need to recreate it on every render loop iteration and the children won't change nearly as frequently.

Reply:
It's not about caching the reversed list, it's about pooling the list objects. And the usage of the list itself is to ensure to not use to much enumerable operators which will create lot of intermediate objects (especially the `OrderBy` which will -most probably- materialize the source anyway).